minject/registry.py

```diff
@@ -238,28 +309,93 @@ def get(
         Returns:
             The requested object or default if not found.
         """
+        if _is_key_async(key):
+            raise AssertionError(
+                "cannot use synchronous get on async object (object marked with @async_context)"
+            )
+
         if key == object:
             return None  # NEVER auto-init plain object
 
         if isinstance(key, str):
             return _unwrap(self._by_name.get(key, RegistryWrapper(cast(T, default))))
 
         meta = _get_meta_from_key(key)
+        maybe_class = self._get_if_already_in_registry(key, meta)
+        if maybe_class is not None:
+            return maybe_class
 
-        if isinstance(key, type):
-            # if a type has metadata attached to it as an attribute,
-            # the registry must use that metadata to construct the object
-            # or query for a constructed object. This is because the user
-            # has intentionally added metadata to the class, and thus
-            # we should not use metadata inherited from interfaces.
-            if _get_meta(key, include_bases=False) is not None:
-                return _unwrap(self._get_by_metadata(meta, default))
+        return _unwrap(self._get_by_metadata(meta, default))
 
-            obj_list = self._by_iface.get(key)
-            if obj_list:
-                return _unwrap(obj_list[0])
+    async def aget(self, key: "RegistryKey[T]") -> Optional[T]:
+        """
+        Resolve objects marked with the @async_context decorator.
+        """
+        if not _is_key_async(key):
+            raise AssertionError("key must be async to use aget")
 
-        return _unwrap(self._get_by_metadata(meta, default))
+        if not self._async_entered:
+            raise AssertionError("cannot use aget outside of async context")
+
+        meta = _get_meta_from_key(key)
+        maybe_class = self._get_if_already_in_registry(key, meta)
+        if maybe_class is not None:
+            return maybe_class
+
+        by_meta = await self._aget_by_metadata(meta)
+        return _unwrap(by_meta)
+
+    def _get_if_already_in_registry(
+        self, key: "RegistryKey[T]", meta: "RegistryMetadata[T]"
+    ) -> Optional[T]:
+        # retrieve the class metdata, and the metadata of class
+        # without inherited metadata
+        meta = _get_meta_from_key(key)
+
+        # if the class has already been registered, return it
+        if meta in self._by_meta:
+            return _unwrap(self._by_meta[meta])
+
+        # following checks only apply if key is a class
+        if not isinstance(key, type):
+            return None
+
+        # If the class has no metadata, but a parent has been registered in registry,
```

Comment:
I'm having trouble understanding what's happening here. It seems like we're explicitly _not_ checking the parents (`include_bases=False`)?

But then we are checking `_by_iface`, which will give us some type that _implements_ (i.e. is a “child” of) `T`. So I think the code is correct but the comment is backwards..?

Reply:
I believe that `_by_interfaces` is populated by `meta.interfaces` which is itself calculated as `[cls for cls in inspect.getmro(cls) if cls is not object]`. Based on this, I believe `_by_interfaces` contains the interfaces `key` implements (I.E. similar to parent classes).

So I think "check parents" is close but not totally right. I've updated the comment to be more explicit about "an object exists in the registry that implements an interface the class implements...".

Agree this is all very tricky. Trying to think of a refactor idea to spin off this conversation. Unsure of one but if I can think of one I'll add an issue.

Reply:
I chatted about this with @bcmills, and my understanding of the `_by_iface` look up is most likely flawed. I was assuming that the `_by_iface` lookup looks for objects THAT implement the interfaces `key` implements, however, we should be looking concrete subtypes of `key`. This makes more sense, as a `key` subtype could be swapped for the `key` type.

I'll update the comment to fix this after I've reread the relevant code (that I previously misread).